internal/zero/controller/usagereporter/usagereporter.go

```diff
@@ -0,0 +1,217 @@
+// Package usagereporter reports usage for a cluster.
+//
+// Usage is determined from session and user records in the databroker. The usage reporter
+// uses SyncLatest and Sync to retrieve this data, builds a collection of records and then
+// sends them to the Zero Cluster API every minute.
+//
+// All usage users are reported on start but only the changed users are reported while running.
+// The Zero Cluster API is tolerant of redundant data.
+package usagereporter
+
+import (
+	"cmp"
+	"context"
+	"sync"
+	"time"
+
+	backoff "github.com/cenkalti/backoff/v4"
+	set "github.com/hashicorp/go-set/v3"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/pomerium/pomerium/internal/log"
+	"github.com/pomerium/pomerium/pkg/cryptutil"
+	"github.com/pomerium/pomerium/pkg/grpc/databroker"
+	"github.com/pomerium/pomerium/pkg/grpc/session"
+	"github.com/pomerium/pomerium/pkg/grpc/user"
+	"github.com/pomerium/pomerium/pkg/zero/cluster"
+)
+
+// API is the part of the Zero Cluster API used to report usage.
+type API interface {
+	ReportUsage(ctx context.Context, req cluster.ReportUsageRequest) error
+}
+
+type usageReporterRecord struct {
+	userID         string
+	userEmail      string
+	lastSignedInAt time.Time
+}
+
+// A UsageReporter reports usage to the zero api.
+type UsageReporter struct {
+	api            API
+	organizationID string
+	reportInterval time.Duration
+
+	mu       sync.Mutex
+	byUserID map[string]usageReporterRecord
+	updates  *set.Set[string]
+}
+
+// New creates a new UsageReporter.
+func New(api API, organizationID string, reportInterval time.Duration) *UsageReporter {
+	return &UsageReporter{
+		api:            api,
+		organizationID: organizationID,
+		reportInterval: reportInterval,
+
+		byUserID: make(map[string]usageReporterRecord),
+		updates:  set.New[string](0),
+	}
+}
+
+// Run runs the usage reporter.
+func (ur *UsageReporter) Run(ctx context.Context, client databroker.DataBrokerServiceClient) error {
+	ctx = log.Ctx(ctx).With().Str("organization-id", ur.organizationID).Logger().WithContext(ctx)
+
+	// first initialize the user collection
+	serverVersion, latestSessionRecordVersion, latestUserRecordVersion, err := ur.runInit(ctx, client)
+	if err != nil {
+		return err
+	}
+
+	// run the continuous sync calls and periodically report usage
+	return ur.runSync(ctx, client, serverVersion, latestSessionRecordVersion, latestUserRecordVersion)
+}
+
+func (ur *UsageReporter) report(ctx context.Context, records []usageReporterRecord) error {
+	req := cluster.ReportUsageRequest{
+		Users: convertUsageReporterRecords(ur.organizationID, records),
+	}
+	return backoff.Retry(func() error {
+		log.Debug(ctx).Int("updated-users", len(req.Users)).Msg("reporting usage")
+		err := ur.api.ReportUsage(ctx, req)
+		if err != nil {
+			log.Warn(ctx).Err(err).Msg("error reporting usage")
+		}
+		return err
+	}, backoff.WithContext(backoff.NewExponentialBackOff(), ctx))
```

Comment:
my reading of this loop is that we capture the updates and try to deliver them for the maximum of default exponential update (15 minutes). 
after that period, the undelivered records are dropped and do not go back into the update set. 

would it make sense to consider moving this retry up to the `runReporter` loop so that undelivered updates would be merged with pending updates or dropping records is intentional?

Reply:
My understanding is that `backoff.Retry` returns an error once stop is reached. This is returned by `report`, which is returned by `runReporter`, which is returned by `runSync`, which is returned by `Run`, which will be retried by `WithBackoff`.

The retry was done here to avoid excessive calls to sync in the databroker. But I can change it to just always return the error if that's preferable.

Reply:
what I meant is maybe combining this retry loop with the `runReporter` main loop, so that unsent records would merge back with `ur.updates` until they can be spooled to the cloud